views.py

```diff
@@ -68,23 +68,27 @@ def download(request, path, rest_call=False, use_async=True, *args, **kwargs):
 
     resource_cls = check_resource_type(res.resource_type)
 
+    if federated_path:
+        res_root = os.path.join(federated_path, res_id)
+    else:
+        res_root = res_id
+
+    bag_modified = "false"
+    metadata_dirty = "false"
+    if istorage.exists(res_root):
+        bag_modified = istorage.getAVU(res_root, 'bag_modified')
+        metadata_dirty = istorage.getAVU(res_root, 'metadata_dirty')
+
     if is_bag_download:
         # do on-demand bag creation
-        bag_modified = "false"
         # needs to check whether res_id collection exists before getting/setting AVU on it
         # to accommodate the case where the very same resource gets deleted by another request
         # when it is getting downloaded
-        if federated_path:
-            res_root = os.path.join(federated_path, res_id)
-        else:
-            res_root = res_id
 
         # send signal for pre_check_bag_flag
         pre_check_bag_flag.send(sender=resource_cls, resource=res)
-
-        if istorage.exists(res_root):
-            bag_modified = istorage.getAVU(res_root, 'bag_modified')
         if bag_modified == "true":
+            create_bag_files(res, fed_zone_home_path=res.resource_federation_path)
```

Comment:
create_bag_files() should only be called when metadata_dirty flag is true here, right? I think another condition is needed here to call create_bag_files() only when metadata_dirty flag is true.